File: C++/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <vector>

using namespace std;

int dfs(const vector<int> &numbers, int target, int idx, int sum) {
  if (idx == numbers.size()) {
    if (target == sum) {
      return 1;
    } else {
      return 0;
    }
  }

  int add_path = dfs(numbers, target, idx + 1, sum + numbers[idx]);
  int sub_path = dfs(numbers, target, idx + 1, sum - numbers[idx]);
  return add_path + sub_path;
}

int solution(vector<int> numbers, int target) {
  int answer = dfs(numbers, target, 0, 0);
  return answer;
}

File: JavaScript/프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
function dfs(numbers, target, idx, sum) {
  if (numbers.length === idx) {
    if (target === sum) {
      return 1;
    } else {
      return 0;
    }
  }

  const add_path = dfs(numbers, target, idx + 1, sum + numbers[idx]);
  const sub_path = dfs(numbers, target, idx + 1, sum - numbers[idx]);
  return add_path + sub_path;
}

function solution(numbers, target) {
  const answer = dfs(numbers, target, 0, 0);
  return answer;
}
